lib/pacemaker/pcmk_sched_promotable.c

```diff
@@ -272,131 +272,217 @@ sort_promotable_instance(gconstpointer a, gconstpointer b, gpointer data_set)
     return sort_clone_instance(a, b, data_set);
 }
 
+/*!
+ * \internal
+ * \brief Applies each child's sort index to the promotable clone
+ *        wrapper's node weights.
+ *
+ * Updates the clone wrapper's node weights based on children's
+ * promotion preference (or promoted role).
+ *
+ * \param[in,out] rsc Clone wrapper.
+ *
+ * \note This function assumes that each child's sort index has already
+ *       been set based on promotion preference and role.
+ */
 static void
-promotion_order(pe_resource_t *rsc, pe_working_set_t *data_set)
+apply_sort_indices_to_wrapper_weights(pe_resource_t *rsc)
 {
-    GList *gIter = NULL;
-    pe_node_t *node = NULL;
-    pe_node_t *chosen = NULL;
-    clone_variant_data_t *clone_data = NULL;
-    char score[33];
-    size_t len = sizeof(score);
-
-    get_clone_variant_data(clone_data, rsc);
-
-    if (clone_data->merged_master_weights) {
-        return;
-    }
-    clone_data->merged_master_weights = TRUE;
-    pe_rsc_trace(rsc, "Merging weights for %s", rsc->id);
-    pe__set_resource_flags(rsc, pe_rsc_merging);
-
-    for (gIter = rsc->children; gIter != NULL; gIter = gIter->next) {
-        pe_resource_t *child = (pe_resource_t *) gIter->data;
-
-        pe_rsc_trace(rsc, "Sort index: %s = %d", child->id, child->sort_index);
-    }
-    pe__show_node_weights(true, rsc, "Before", rsc->allowed_nodes, data_set);
-
-    gIter = rsc->children;
-    for (; gIter != NULL; gIter = gIter->next) {
+    for (GList *gIter = rsc->children; gIter != NULL; gIter = gIter->next) {
         pe_resource_t *child = (pe_resource_t *) gIter->data;
+        pe_node_t *chosen = child->fns->location(child, NULL, 0);
+        pe_node_t *node = NULL;
+        char score[33];
+        size_t len = sizeof(score);
```

Comment:
score and len should be outside the for loop, the idea is to avoid shuffling memory around. I'm sure compilers are smart enough to do that anyway, but I think it's worth making the intent clear

BTW 33 is totally bogus, someone thought 32-bit integers needed 32 characters plus a nul. score2char_stack() truncates to +/-INFINITY anyway so 10 characters should be sufficient ("-INFINITY")

Reply:
> score and len should be outside the for loop, the idea is to avoid shuffling memory around.

Yeah, thanks for catching that. No need to re-allocate an array on every iteration.

> BTW 33 is totally bogus

I wondered what that was about.

Reply:
Fixed